workflows/mapping/Snakefile

```diff
@@ -249,4 +257,67 @@ rule merge_bam:
     wrapper:
         wrapper_for('samtools/merge')
 
-# vim: ft=python
+
+rule fastq_count:
+    input: '{prefix}.fastq.gz'
+    output: '{prefix}.fastq.count'
+    shell: "gunzip -c {input[0]} | echo $((`wc -l` / 4)) > {output[0]}"
+
+
+rule bam_count:
+    input: '{prefix}.bam'
+    output: '{prefix}.bam.count'
+    shell: "samtools view -c {input[0]} > {output[0]}"
+
+rule sample_count_table:
+    input: fq = '{prefix}.fastq.count',
+           cutadapt = '{prefix}.cutadapt.fastq.count',
+           align = '{prefix}.cutadapt.ht2.bam.count',
+           dedup = '{prefix}.cutadapt.ht2.unique.sort.dedup.bam.count'
+    output: '{prefix}_counts.csv',
+    run:
+        import pandas as pd
+        def read_count(count):
+            with open(count, 'r') as FH:
+                return int(FH.read().strip())
+        
+        fq = read_count(input.fq)
+        cutadapt = read_count(input.cutadapt)
+        align = read_count(input.align)
+        dedup = read_count(input.dedup)
+
+        df = pd.DataFrame([[wildcards.prefix, fq, cutadapt, align, dedup]], columns=['sampleID', 'Number Reads', 'Number Trimmed Reads', 'Number Aligned', 'Number Uniquely Aligned'])
+        df.to_csv(output[0], index=False)
+
+rule count_table:
+    input: SH.build_targets(['{runLevel}_counts.csv'])
+    output: csv = 'count_table.csv',
+            html = 'count_table.html'
+    run:
+        import pandas as pd
+
+        dfs = [pd.read_csv(x) for x in input]
+        df = pd.concat(dfs)
+        df.to_csv(output.csv, index=False)
+        with open(output.html, 'w') as OUT:
+            df.to_html(buf=OUT, index=False, classes='table table-striped')
+
+rule report:
+    input: Table1='count_table.html'
+    output: html='report.html'
+    run:
+        from reporting import report
+
+        myIn = {}
+        with open(input.Table1, 'r') as FH:
+            myIn['Table1'] = FH.read().strip()
+        
```

Comment:
I was initially thinking of one giant report with conditionals to detect the presence of a rule's set of chunks, and if it's there then iterate over them. I didn't fill out that example enough to support that though (I guess chunk.id was ambiguous. Would `id` be rulename or samplename?). But one jinja template per rule could work too.

I guess the question is, what's easiest to maintain and edit? In a year from now when I want to edit the description of the bowtie2 part of the report, I can either edit the section nested in the conditional in the report, or if the jinja template for this chunk is in the `bowtie2_chunk` rule itself I wouldn't have to track down an additional file.

The pickling thing seems weird to me. If we store the output as `temp("bowtie2_chunk.html")` or `temp("chunks/bowtie2_chunk.html")` the final report should be able to figure out what the args to `lcdb.Chunk()` should be, so the rules become:

```python
import os
import pandas as pd
import lcdb.odo
from lcdb.reporting import Chunk

rule bowtie2_chunk:
    input: _bt2_input_func      # provide list of all *.bam.log
    output: temp('bowtie2_chunk.html')
    run:
         dfs = []
         for i in input:
              dfs.append(odo(i))
         df = pd.concat(dfs)
         with open(output[0], 'w' ) as fout:
             fout.write(df_to_html(classes='table stripped-table'))

rule final_report:
    input: _chunks  #  list of HTML files
    output: 'report.html'
    run:
        chunks = [
            Chunk(os.path.basename(x).replace('_chunk.html', ''), x)
            for x in input]

        #...get base.html template as tp, then
        tp.render(chunks=chunks)
```

A nice side-effect of this is that each chunk is a standalone HTML that could be distributed separately if needed.